rllib/core/models/tests/test_base_models.py

```diff
@@ -0,0 +1,267 @@
+                    # to distinguish between frameworks here.
+                    return TestModel(self)
+
+            # 1) Check if model behaves correctly with always_check_shapes=True first.
+            # We expect model to raise an error if the output shapes are not correct.
+            # This is the behaviour we use for debugging with model specs.
+
+            config = CatModelConfig(always_check_shapes=True)
+
+            model = config.build(framework="spam")
+
+            # We want to raise an output spec validation error here since the output
+            # has the wrong shape
+            with self.assertRaisesRegex(ValueError, "output spec validation failed"):
+                model({"in": torch.Tensor([[1]])})
+
+            # 2) Check if model behaves correctly with always_check_shapes=False.
+            # We don't expect model to raise an error.
+            # This is the more performant default behaviour
+
+            config = CatModelConfig(always_check_shapes=False)
+
+            model = config.build(framework="spam")
+
+            model({"in_1": [[1]]})
+
+    @unittest.skipIf(not _dynamo_is_available(), "torch._dynamo not available")
+    def test_torch_compile_with_model(self):
+        """Tests if torch.compile() does not encounter any breaks.
+
+        torch.compile() should not encounter any breaks when model is on its
+        code path by default. This test checks if this is the case.
+        """
+
+        class SomeTorchModel(TorchModel):
+            """Simple model that produces bad outputs."""
+
+            def __init__(self, config):
+                super().__init__(config)
+                self._model = torch.nn.Linear(1, 1)
+
+            def get_output_specs(self):
+                return SpecDict(
+                    {
+                        "out": TensorSpec("b, h", h=1, framework="torch"),
+                    }
+                )
+
+            def get_input_specs(self):
+                return SpecDict(
+                    {
+                        "in": TensorSpec("b, h", h=1, framework="torch"),
+                    }
+                )
+
+            def _forward(self, input_dict):
+                return {"out": self._model(input_dict["in"])}
+
+        @dataclass
+        class SomeTorchModelConfig(ModelConfig):
+            def build(self, framework: str):
+                return SomeTorchModel(self)
+
+        config = SomeTorchModelConfig()
+
+        model = config.build(framework="spam")
+
+        # This could be the forward method of an RL Module that we torch compile
+        def compile_me(input_dict):
+            return model(input_dict)
+
+        import torch._dynamo as dynamo
+
+        dynamo_explanation = dynamo.explain(compile_me, {"in": torch.Tensor([[1]])})
+
+        # There should be only one break reason - `return_value` - since inputs and
+        # outputs are not checked
+        break_reasons_list = dynamo_explanation[4]
```

Comment:
I see. can you explain what index 4 is? and reference some documentation on the web :)

Reply:
Yeah that's the funny part.
For now, torch._dynamo.explain does not even have a docstring.
We'll need to rivisit this test for sure!
I'd suggest we wait until torch 2.x has stabilized a little in that sense.